File: packages/contracts/src/archive.ts
import { ContractInputError, childPointer } from "./canonical-json.js";

export interface CanonicalArchiveFile {
  contentBase64: string;
  kind: "file";
  path: string;
}

export interface CanonicalArchiveDirectory {
  kind: "directory";
  path: string;
}

export type CanonicalArchiveEntry = CanonicalArchiveDirectory | CanonicalArchiveFile;

export interface CanonicalArchiveInput {
  contractId: "canonical-archive";
  entries: CanonicalArchiveEntry[];
  schemaVersion: 1;
}

interface NormalizedEntry {
  content: Buffer;
  inputIndex: number;
  kind: "directory" | "file";
  path: string;
}

const BLOCK_SIZE = 512;

function invalidPath(pointer: string, message: string): never {
  throw new ContractInputError("unsafe_path", pointer, message);
}

function normalizeEntry(entry: CanonicalArchiveEntry, index: number): NormalizedEntry {
  const pointer = childPointer(childPointer("", "entries"), index);
  const pathPointer = childPointer(pointer, "path");
  const normalized = entry.path.normalize("NFC");
  const parts = normalized.split("/");

  if (
    normalized !== entry.path ||
    normalized.startsWith("/") ||
    normalized.includes("\\") ||
    normalized.includes("\0") ||
    parts.some((part) => part === "" || part === "." || part === "..")
  ) {
    invalidPath(pathPointer, `Unsafe archive path: ${entry.path}`);
  }

  const archivePath = entry.kind === "directory" ? `${normalized}/` : normalized;
  if (Buffer.byteLength(archivePath, "utf8") > 100) {
    invalidPath(pathPointer, "Canonical ustar paths are limited to 100 UTF-8 bytes.");
  }

  let content = Buffer.alloc(0);
  if (entry.kind === "file") {
    content = Buffer.from(entry.contentBase64, "base64");
    if (content.toString("base64") !== entry.contentBase64) {
      throw new ContractInputError(
        "format",
        childPointer(pointer, "contentBase64"),
        "Invalid canonical base64.",
      );
    }
  }

  return { content, inputIndex: index, kind: entry.kind, path: archivePath };
}

export function normalizeArchiveEntries(entries: CanonicalArchiveEntry[]): NormalizedEntry[] {
  const normalized = entries.map(normalizeEntry);
  normalized.sort((left, right) => Buffer.compare(Buffer.from(left.path), Buffer.from(right.path)));

  const collision = (entry: NormalizedEntry): never =>
    invalidPath(
      childPointer(childPointer(childPointer("", "entries"), entry.inputIndex), "path"),
      `Colliding archive path: ${entry.path}`,
    );
  for (const [index, entry] of normalized.entries()) {
    const entryBase = entry.path.replace(/\/$/, "").toLowerCase();
    for (const prior of normalized.slice(0, index)) {
      const priorBase = prior.path.replace(/\/$/, "").toLowerCase();
      if (entryBase === priorBase) {
        collision(entry.inputIndex > prior.inputIndex ? entry : prior);
      }
      if (entryBase.startsWith(`${priorBase}/`) && prior.kind === "file") {
        collision(entry.inputIndex > prior.inputIndex ? entry : prior);
      }
      if (priorBase.startsWith(`${entryBase}/`) && entry.kind === "file") {
        collision(entry.inputIndex > prior.inputIndex ? entry : prior);
      }
    }
  }
  return normalized;
}

function writeAscii(target: Buffer, offset: number, width: number, value: string): void {
  const bytes = Buffer.from(value, "ascii");
  if (bytes.length > width) {
    throw new ContractInputError("range", "", `ustar field exceeds ${width} bytes.`);
  }
  bytes.copy(target, offset);
}

function octal(value: number, width: number): string {
  const digits = value.toString(8);
  if (digits.length > width - 1) {
    throw new ContractInputError("range", "", `Value ${value} exceeds ustar field width ${width}.`);
  }
  return `${digits.padStart(width - 1, "0")}\0`;
}

function headerFor(entry: NormalizedEntry): Buffer {
  const header = Buffer.alloc(BLOCK_SIZE);
  Buffer.from(entry.path, "utf8").copy(header, 0);
  writeAscii(header, 100, 8, octal(entry.kind === "directory" ? 0o755 : 0o644, 8));
  writeAscii(header, 108, 8, octal(0, 8));
  writeAscii(header, 116, 8, octal(0, 8));
  writeAscii(header, 124, 12, octal(entry.content.length, 12));
  writeAscii(header, 136, 12, octal(0, 12));
  header.fill(0x20, 148, 156);
  writeAscii(header, 156, 1, entry.kind === "directory" ? "5" : "0");
  writeAscii(header, 257, 6, "ustar\0");
  writeAscii(header, 263, 2, "00");
  writeAscii(header, 329, 8, octal(0, 8));
  writeAscii(header, 337, 8, octal(0, 8));

  const checksum = header.reduce((sum, byte) => sum + byte, 0);
  writeAscii(header, 148, 8, `${checksum.toString(8).padStart(6, "0")}\0 `);
  return header;
}

export function canonicalArchiveBytes(value: unknown): Buffer {
  if (
    value === null ||
    typeof value !== "object" ||
    !("entries" in value) ||
    !Array.isArray(value.entries)
  ) {
    throw new ContractInputError(
      "type",
      "/entries",
      "Canonical archive input requires an entries array.",
    );
  }

  const entries = normalizeArchiveEntries(value.entries as CanonicalArchiveEntry[]);
  const chunks: Buffer[] = [];
  for (const entry of entries) {
    chunks.push(headerFor(entry));
    if (entry.kind === "file") {
      chunks.push(entry.content);
      const padding = (BLOCK_SIZE - (entry.content.length % BLOCK_SIZE)) % BLOCK_SIZE;
      if (padding > 0) {
        chunks.push(Buffer.alloc(padding));
      }
    }
  }
  chunks.push(Buffer.alloc(BLOCK_SIZE * 2));
  return Buffer.concat(chunks);
}

File: python/src/palimpsest/contracts/archive.py
from __future__ import annotations

import base64
from dataclasses import dataclass
from typing import Any
from unicodedata import normalize

from .canonical_json import ContractInputError, child_pointer

BLOCK_SIZE = 512


@dataclass(frozen=True)
class _Entry:
    path: str
    kind: str
    content: bytes
    input_index: int


def _invalid_path(pointer: str, message: str) -> None:
    raise ContractInputError("unsafe_path", pointer, message)


def _normalize_entry(entry: dict[str, Any], index: int) -> _Entry:
    pointer = child_pointer(child_pointer("", "entries"), index)
    path_pointer = child_pointer(pointer, "path")
    path = entry.get("path")
    kind = entry.get("kind")
    if not isinstance(path, str) or kind not in {"file", "directory"}:
        raise ContractInputError("type", pointer, "Archive entry has an invalid path or kind.")
    normalized = normalize("NFC", path)
    parts = normalized.split("/")
    if (
        normalized != path
        or normalized.startswith("/")
        or "\\" in normalized
        or "\0" in normalized
        or any(part in {"", ".", ".."} for part in parts)
    ):
        _invalid_path(path_pointer, f"Unsafe archive path: {path}")
    archive_path = f"{normalized}/" if kind == "directory" else normalized
    if len(archive_path.encode("utf-8")) > 100:
        _invalid_path(path_pointer, "Canonical ustar paths are limited to 100 UTF-8 bytes.")

    content = b""
    if kind == "file":
        encoded = entry.get("contentBase64")
        if not isinstance(encoded, str):
            raise ContractInputError(
                "required",
                child_pointer(pointer, "contentBase64"),
                "File content is required.",
            )
        try:
            content = base64.b64decode(encoded, validate=True)
        except ValueError as error:
            raise ContractInputError(
                "format",
                child_pointer(pointer, "contentBase64"),
                "Invalid canonical base64.",
            ) from error
        if base64.b64encode(content).decode("ascii") != encoded:
            raise ContractInputError(
                "format",
                child_pointer(pointer, "contentBase64"),
                "Invalid canonical base64.",
            )
    return _Entry(path=archive_path, kind=kind, content=content, input_index=index)


def normalize_archive_entries(entries: list[dict[str, Any]]) -> list[_Entry]:
    normalized = [_normalize_entry(entry, index) for index, entry in enumerate(entries)]
    normalized.sort(key=lambda entry: entry.path.encode("utf-8"))
    for index, entry in enumerate(normalized):
        entry_base = entry.path.removesuffix("/").lower()
        for prior in normalized[:index]:
            prior_base = prior.path.removesuffix("/").lower()
            if (
                entry_base == prior_base
                or (entry_base.startswith(f"{prior_base}/") and prior.kind == "file")
                or (prior_base.startswith(f"{entry_base}/") and entry.kind == "file")
            ):
                target = entry if entry.input_index > prior.input_index else prior
                _invalid_path(
                    f"/entries/{target.input_index}/path",
                    f"Colliding archive path: {target.path}",
                )
    return normalized


def _octal(value: int, width: int) -> bytes:
    digits = format(value, "o")
    if len(digits) > width - 1:
        raise ContractInputError("range", "", f"Value {value} exceeds ustar field width {width}.")
    return f"{digits:0>{width - 1}}\0".encode("ascii")


def _header(entry: _Entry) -> bytes:
    header = bytearray(BLOCK_SIZE)
    path = entry.path.encode("utf-8")
    header[0 : len(path)] = path
    header[100:108] = _octal(0o755 if entry.kind == "directory" else 0o644, 8)
    header[108:116] = _octal(0, 8)
    header[116:124] = _octal(0, 8)
    header[124:136] = _octal(len(entry.content), 12)
    header[136:148] = _octal(0, 12)
    header[148:156] = b" " * 8
    header[156:157] = b"5" if entry.kind == "directory" else b"0"
    header[257:263] = b"ustar\0"
    header[263:265] = b"00"
    header[329:337] = _octal(0, 8)
    header[337:345] = _octal(0, 8)
    checksum = sum(header)
    header[148:156] = f"{checksum:06o}\0 ".encode("ascii")
    return bytes(header)


def canonical_archive_bytes(value: Any) -> bytes:
    if not isinstance(value, dict) or not isinstance(value.get("entries"), list):
        raise ContractInputError(
            "type",
            "/entries",
            "Canonical archive input requires an entries array.",
        )
    entries = normalize_archive_entries(value["entries"])
    chunks: list[bytes] = []
    for entry in entries:
        chunks.append(_header(entry))
        if entry.kind == "file":
            chunks.append(entry.content)
            padding = (-len(entry.content)) % BLOCK_SIZE
            if padding:
                chunks.append(bytes(padding))
    chunks.append(bytes(BLOCK_SIZE * 2))
    return b"".join(chunks)
